libraries/AP_BattMonitor/AP_BattMonitor_FuelLevel_Analog.cpp

```diff
@@ -0,0 +1,62 @@
+/*
+ * This file is free software: you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License as published by the
+ * Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This file is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * Code by Charlie Johnson
+ */
+
+#include <AP_HAL/AP_HAL.h>
+#include "AP_BattMonitor_FuelLevel_Analog.h"
+
+extern const AP_HAL::HAL& hal;
+
+// Constructor
+AP_BattMonitor_FuelLevel_Analog::AP_BattMonitor_FuelLevel_Analog(AP_BattMonitor &mon, AP_BattMonitor::BattMonitor_State &mon_state, AP_BattMonitor_Params &params) :
+    AP_BattMonitor_Analog(mon, mon_state, params)
+{
+    _state.voltage = 1.0; // show a fixed voltage of 1v
+
+    _analog_source = hal.analogin->channel(_curr_pin);
+
+    // need to add check
+    _state.healthy = false;
+}
+
+/*
+  read - read the "voltage" and "current"
+*/
+void AP_BattMonitor_FuelLevel_Analog::read()
+{
+    if (!_analog_source->set_pin(_curr_pin)) {
+        _state.healthy = false;
+        return;
+    }
+
+    float voltage = _analog_source->voltage_average();
+    uint32_t now_us = AP_HAL::micros();
+
+    /*
+      this driver assumes that CAPACITY is set to tank volume in millilitres.
+     */
+    float tank_level_ratio = voltage / 3.3;
```

Comment:
Pretty sure we need a scalar ratio for these, as there may be external dividers in front of this that we can't control, and to correct for scaling errors from divider components.

Reply:
Also needs to cope with an input that works inside of the not full range (for example I've got a sensor on the way that has 0.25 - 4.75V output for the fuel level, which I can use an ADC to cut down to the right voltage range, but I need to be able to chop off the bottom bit of the range or it will never detect an empty tank correctly.

Reply:
My other thought was to use read_average to skip the pin_scalar.

Reply:
You don't want to just deal with the internal scalers, there's also external scalars that we might have to add, and we won't be able to perfectly map to the 0 - 3.3V range, especially not if we want some error tolerance.

Reply:
A couple new parameters for low/high voltage and linear interpolation? (My end goal with all of this is to get a CAN node to output FuelTankStatus, but need the ground work first.)

Reply:
That sounds perfect to me. That should solve the ratio thing I pointed out above with fewer parameters as well. It also makes it easy to calibrate because you can just fill it all the way, and drain it.

Maybe actually just stick the current ADC pin reading as the voltage because it makes easy debug?